velox/vector/DecodedVector.cpp

```diff
@@ -383,16 +383,28 @@ DecodedVector::DictionaryWrapping DecodedVector::dictionaryWrapping(
   BufferPtr indices = copyIndicesBuffer(indices_, size, wrapper.pool());
   // Only copy nulls if we have nulls coming from one of the wrappers, don't
   // do it if nulls are missing or from the base vector.
-  // TODO: remove the check for hasExtraNulls_ after #3553 is merged.
-  BufferPtr nulls =
-      hasExtraNulls_ ? copyNullsBuffer(nulls_, size, wrapper.pool()) : nullptr;
+  BufferPtr nulls;
+  if (hasExtraNulls_) {
+    nulls = copyNullsBuffer(nulls_, size, wrapper.pool());
+  } else {
+    if (!wrapper.mayHaveNulls()) {
+      nulls =
+          AlignedBuffer::allocate<bool>(size, wrapper.pool(), bits::kNotNull);
+
+      auto* mutableNulls = nulls->asMutable<uint64_t>();
+
+      for (auto i = 0; i < size; ++i) {
+        if (wrapper.isNullAt(i)) {
```

Comment:
This might cause a lot of confusion, the wrapper that you are passing here is no longer the original vector which should contain the dictionary wraps, since now you are passing the base directly in L420.

Also, here we are directly mapping top level row i ->  base row i , where as we need to translate the null rows of the base via the wrappings. You need to do something similar to what we do here: DecodedVector::setFlatNulls().

My recommendation would be to get rid of 'wrapper' input here (if you cannot, then we need to keep it in DecodedVector::wrap as well) and finally use the logic similar to what we do in DecodedVector::setFlatNulls() to translate the base's null rows to top level rows

Reply:
I assume you refer to :
```
 applyToRows(rows, [&](vector_size_t row) {
      if (!bits::isBitNull(nulls_, row) &&
          (leafNulls && bits::isBitNull(leafNulls, indices_[row]))) {
        bits::setNull(copiedNulls, row);
      }
    });
  ```
I think its another bug that we do that for only selected rows. right?
how is that different than the code here otherwise?

Reply:
since now you are passing the base directly in L420.
 
 not sure I get this part.

Reply:
```
My recommendation would be to get rid of 'wrapper' input here (if you cannot, then we need to keep it in DecodedVector::wrap as well) and finally use the logic similar to what we do in DecodedVector::setFlatNulls() to translate the base's null rows to top level rows
```
I can do that I was planning to do it as part of a different diff.